src/i18n.ts

```diff
@@ -0,0 +1,248 @@
+import { Context, Fluent, FluentType, resolve } from "./deps.ts";
+
+import type {
+  FluentBundleOptions,
+  FluentValue,
+  LocaleId,
+  Middleware,
+  NextFunction,
+  Scope,
+  TranslationContext,
+} from "./deps.ts";
+
+import { exists, readLocalesDir } from "./utils.ts";
+
+import type {
+  I18nConfig,
+  I18nContextFlavor,
+  TranslateFunction,
+} from "./types.ts";
+
+class FluentContext extends FluentType<string> {
+  constructor(value: string, private defaultValue = "") {
+    super(value);
+  }
+
+  toString(scope: Scope): string {
+    let ctx = JSON.parse(scope.args?.ctx as string);
+    const keys = this.value.split(".");
+    for (const key of keys) {
+      ctx = ctx[key];
+      if (ctx === undefined) break;
+    }
+    return ctx ?? this.defaultValue;
+  }
+}
+
+/**
+ * A custom function for accessing `Context` properties directly from
+ * Fluent translation files.
+ */
+export function CTX(args: FluentValue[]) {
+  if (args[1] !== undefined && typeof args[1] !== "string") {
+    throw new TypeError("CTX: Only string type is allowed for defaultValue.");
+  }
+
+  const key = args[0];
+  const defaultValue = typeof args[1] === "string" ? args[1] : undefined;
+
+  if (typeof key === "string") {
+    return new FluentContext(key, defaultValue);
+  }
+
+  throw new TypeError(
+    "CTX: Invalid argument type. Both key and defaultValue should be the type of string.",
+  );
+}
+
+export class I18n<C extends Context = Context> {
+  private config: I18nConfig<C>;
+  readonly fluent: Fluent;
+  readonly locales = new Array<string>();
+
+  constructor(config: Partial<I18nConfig>) {
+    this.config = {
+      defaultLocale: "en",
+      fluentBundleOptions: { functions: {} },
+      ...config,
+    };
+
+    if (
+      this.config.fluentBundleOptions &&
+      this.config.fluentBundleOptions.functions
+    ) {
+      this.config.fluentBundleOptions.functions["CTX"] = CTX;
+    } else {
+      this.config.fluentBundleOptions = {
+        ...this.config.fluentBundleOptions,
+        functions: { CTX },
+      };
+    }
+
+    this.fluent = new Fluent(this.config.fluentOptions);
+  }
+
+  /**
+   * Load and register translations from a directory to the Fluent instance.
+   * @param directory Path to the directory to look for the translation files.
+   * Translation files in that directory should end with `.ftl` extension.
+   */
+  async loadLocalesDir(directory: string): Promise<void> {
```

Comment:
About the sync version, we have to talk with @slavafomin  if he is interested in updating [@the-moebius/fluent](https://github.com/the-moebius/fluent)'s [`addTranslation`](https://github.com/the-moebius/fluent/blob/83274d106b07770a5fbb16727f6da3e11aae1edc/src/fluent.ts#L59) (-> [`handleSources`](https://github.com/the-moebius/fluent/blob/83274d106b07770a5fbb16727f6da3e11aae1edc/src/fluent.ts#L69) -> [`readFile`](https://github.com/the-moebius/fluent/blob/83274d106b07770a5fbb16727f6da3e11aae1edc/src/read-file.ts#L9)) to a sync version. On the Deno side, I have already [released](https://github.com/dcdunkan/deno_fluent/commit/689aafab882488e7ed47c73b8352e4895798b4da) a version of better_fluent which doesn't use promises to read files (add translations).

If we have a sync version, we can also bring back the `directory` option in config. It was removed due to not having a sync version of `loadLocalesDir`.

Reply:
Maybe we can also turn this into an issue. I don't really mind if things are async, but it's actually commonly seen that one-time setup IO is sync. That way, it can be used on the top-level and from constructors. It might be quite convenient.

Reply:
Same as https://github.com/grammyjs/i18n/pull/20#discussion_r912065382.

Reply:
I guess this has been resolved in f42daac